### PR TITLE
Record chords

### LINKED ISSUE #4807: Record chords
Edited by @tresf 

In the Piano Roll editor, the "chords" drop-down does not affect notes played through a keyboard, only those placed with a mouse, also impacting recording.

image

---

Original

When the piano-roll *chord-selector* has any other setting than *no-chords*, using  *record* ought to insert the selected* chord* in piano-roll. For now only the pressed note is inserted.

### PR DESCRIPTION
**Pull request by alxdttn**
Checks for currently selected chord when recording from keyboard.

Closes #4807

### CODE REVIEW

File: src/gui/editors/PianoRoll.cpp [hunk 1]
```diff
@@ -3522,6 +3522,23 @@ void PianoRoll::finishRecordNote(const Note & n )
 						it->key(), it->getVolume(),
 						it->getPanning() );
 				n1.quantizeLength( quantization() );
+
+				//Get selected chordModel
+				const InstrumentFunctionNoteStacking::Chord & chord = InstrumentFunctionNoteStacking::ChordTable::getInstance()
+						.getChordByName( m_chordModel.currentText() );
+
+					if( ! chord.isEmpty() )
```
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp]: It'd be better to unindent the `if` block one level.

File: src/gui/editors/PianoRoll.cpp [hunk 2]
```diff
@@ -3522,6 +3522,23 @@ void PianoRoll::finishRecordNote(const Note & n )
...
+					{
+						// if a chord is selected, create following notes in chord mode
+						for( int i = 1; i < chord.size(); i++ )
+						{
+							Note new_note( n.length(), it->pos(), it->key() + chord[i] );
```
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp]: We prefer `camelCase` to `snake_case`, though it's not in the coding convention.

File: src/gui/editors/PianoRoll.cpp [hunk 3]
```diff
@@ -3522,6 +3522,23 @@ void PianoRoll::finishRecordNote(const Note & n )
 						it->key(), it->getVolume(),
 						it->getPanning() );
 				n1.quantizeLength( quantization() );
+
+				//Get selected chordModel
```
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp]: `// get selected chord` is enough, but you can leave this as-is.

File: src/gui/editors/PianoRoll.cpp (hunk 1 above)
Comment (PhysSong) [reply src/gui/editors/PianoRoll.cpp]: Also, `if( a )` is an old style. According to [our coding convention](https://github.com/LMMS/lmms/wiki/Coding-conventions), `if (a)` is correct. However, you can leave this as-is if the change breaks consistency.

File: src/gui/editors/PianoRoll.cpp [hunk 4]
```diff
@@ -3522,6 +3522,23 @@ void PianoRoll::finishRecordNote(const Note & n )
...
+				const InstrumentFunctionNoteStacking::Chord & chord = InstrumentFunctionNoteStacking::ChordTable::getInstance()
+						.getChordByName( m_chordModel.currentText() );
+
+					if( ! chord.isEmpty() )
+					{
+						// if a chord is selected, create following notes in chord mode
+						for( int i = 1; i < chord.size(); i++ )
```
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp]: Here as well.

**Comment by PhysSong** [comment]
Thank you for your contribution! I left some comments regarding coding styles.
Also, please consider creating a new branch when you create a pull request next time.

**Comment by SecondFlight** [comment]
@PhysSong That's on me, they asked in #devtalk if they should create a new branch and I said they didn't need to.

**Comment by zonkmachine** [comment]
Tested. This fixes the issue as reported in #4807. I'm a bit confused as to how the chord function is intended to work though. Even though the chord is recorded only the base note is played back while recording/playing.

**Comment by alxdttn** [comment]
I thought that would be best, as it matches the behavior of using the mouse to place notes when in chord mode, which also seems to only play the bass note.

**Comment by zonkmachine** [comment]
> I thought that would be best, as it matches the behavior of using the mouse to place notes when in chord mode, which also seems to only play the bass note.

Yes. Its enough for this PR. This is good to merge I think and I'll open a new issue for the _'only base note sounds while playing chord notes'_ issue.

**Comment by zonkmachine** [comment]
@PhysSong @alxdttn Merge?

**Pull request merged**